homeassistant/components/utility_meter/sensor.py

```diff
@@ -387,9 +387,17 @@ def _validate_state(state: State | None) -> Decimal | None:
         )
 
     def calculate_adjustment(
-        self, old_state: State | None, new_state_val: Decimal
+        self, old_state: State | None, new_state: State
     ) -> Decimal | None:
-        """Calculate the adjustment based on the old and (assumed valid) new state."""
+        """Calculate the adjustment based on the old and new state."""
+
+        # First check if the new_state is valid (see discussion in PR #88446)
+        try:
+            if (new_state_val := self._validate_state(new_state)) is None:
+                return None
+        except DecimalException as err:
```

Comment:
Why not validate that it can be cast to Decimal in _validate_state (Returning None if not valid)

Reply:
I chose to not do the Exception handling inside `_validate_state` because the error message you want to log depends on the state you are validating, If the new state is invalid, you may want to indicate the state is invalid, while if the new_state is valid but the old one is invalid (and needed), than you want to log both states and show that it cannot be used.

If the logger is already inside `_validate_state`, you will first receive an error that the state is invalid, if the new state is invalid, and receive the same error in the next run, because then, the old state is still invalid and will trigger a new error log.

Another option I'm thinking about is to do the Exception handling inside the validate_state method and then if it fails to parse the Decimal, we return None and throw an appropriate error log in the scope we called the method, i.e. something like:
```python
if (new_state_val := self._validate_state(new_state)) is None:
    _LOGGER.error("Invalid state %s", new_state.state)
   return
```

Do you agree with this solution?

Reply:
That is what I was proposing :)

Reply:
Great, will change it! :)